coderd/insights.go

```diff
@@ -292,6 +292,68 @@ func (api *API) insightsUserLatency(rw http.ResponseWriter, r *http.Request) {
 	httpapi.Write(ctx, rw, http.StatusOK, resp)
 }
 
+// @Summary Get insights about user status counts over time
+// @ID get-insights-about-user-status-counts-over-time
+// @Security CoderSessionToken
+// @Produce json
+// @Tags Insights
+// @Param tz_offset query int true "Time-zone offset (e.g. -2)"
+// @Success 200 {object} codersdk.GetUserStatusCountsOverTimeResponse
+// @Router /insights/user-status-counts-over-time [get]
+func (api *API) insightsUserStatusCountsOverTime(rw http.ResponseWriter, r *http.Request) {
+	ctx := r.Context()
+
+	p := httpapi.NewQueryParamParser()
+	vals := r.URL.Query()
+	tzOffset := p.Int(vals, 0, "tz_offset")
+	p.ErrorExcessParams(vals)
+
+	if len(p.Errors) > 0 {
+		httpapi.Write(ctx, rw, http.StatusBadRequest, codersdk.Response{
+			Message:     "Query parameters have invalid values.",
+			Validations: p.Errors,
+		})
+		return
+	}
+
+	loc := time.FixedZone("", tzOffset*3600)
+	// If the time is 14:01 or 14:31, we still want to include all the
+	// data between 14:00 and 15:00. Our rollups buckets are 30 minutes
+	// so this works nicely. It works just as well for 23:59 as well.
+	nextHourInLoc := time.Now().In(loc).Truncate(time.Hour).Add(time.Hour)
+	// Always return 60 days of data (2 months).
+	sixtyDaysAgo := nextHourInLoc.In(loc).Truncate(24*time.Hour).AddDate(0, 0, -60)
+
+	rows, err := api.Database.GetUserStatusCountsOverTime(ctx, database.GetUserStatusCountsOverTimeParams{
+		StartTime: sixtyDaysAgo,
+		EndTime:   nextHourInLoc,
+	})
+	if err != nil {
+		httpapi.Write(ctx, rw, http.StatusInternalServerError, codersdk.Response{
+			Message: "Internal error fetching user status counts over time.",
+			Detail:  err.Error(),
+		})
+		return
+	}
```

Comment:
Might want to add a check for an unauthorized call, and return a Forbidden vs a 500.

```

```
```suggestion
	if err != nil {
	       if httpapi.IsUnauthorizedError(err) {
		    httpapi.Forbidden(rw)
		    return
		}
		httpapi.Write(ctx, rw, http.StatusInternalServerError, codersdk.Response{
			Message: "Internal error fetching user status counts over time.",
			Detail:  err.Error(),
		})
		return
	}
```

Reply:
What about err no rows? Is it a possible state?

Reply:
Not for `:many` queries. It'll just return an empty slice.

Reply:
This `:many` query will always return at least two rows. one for `@start_time` and one for `@end_time`. And even if that changes, an empty slice is fine too.

Reply:
Adding the unauthorised check though